### PR TITLE
Introduction of tesla hardware in domoticz

### PR DESCRIPTION
**Pull request by MrHobbes74**
I created a domoticz native Tesla hardware module.
The module connects to a Tesla, and the first version provides:
- a battery level indicator
- a alert indicator giving you the status of the car 
(offline, not home, home, cable disconnected, charging, charging stopped)
- a temperature meter indicating the temperature in the tesla
- a switch to turn on / off charging
- a switch to turn on / off climate control
- a switch to turn on / off defrost mode

For safety, switches cannot be set when the car is not at home/around the house.

### CODE REVIEW

File: hardware/Tesla.cpp [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+{
+	int sec_counter = 593;
+	int sec_counter_after_active = 0;
+	bool waking_up = false;
+	int interval = 1000;
+	Log(LOG_STATUS, "Worker started...");
+
+	while (!IsStopRequested(interval))
+	{
+		if (IsStopRequested(0))
```
Comment (gizmocuz) [inline hardware/Tesla.cpp]: The second IsStopRequested is not needed.
When it's time to quit the worker, it stops directly

File: hardware/Tesla.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include "DomoticzHardware.h"
+#include "TeslaApi.h"
+#include "../main/json_helper.h"
```
Comment (gizmocuz) [inline hardware/Tesla.h]: Not sure if you need this include here... (json_helper.h)

**Review by gizmocuz — COMMENTED** [review]
Thanks for this hardware addon ! Some minor questions/comments

File: hardware/Tesla.h (hunk 2 above)
Comment (MrHobbes74) [reply hardware/Tesla.h]: indeed, not needed. Remainder of a previous change. Will remove.

**Comment by MrHobbes74** [comment]
Note that I have separated the API from the worker. Except the result parameters (in json) the tesla details are all in TeslaApi. A next version I plan to fully decouple the worker from the car specifics, such that the worker can become an Electrical Vehicle, that can support various API's.

**Comment by gizmocuz** [comment]
Thanks for the changes !!!
Just a question before i merge your code.... ;) ... do you plan to create more source files when you are going to decouple ?
If yes, maybe you could move all into a Tesla folder ? Keeps everything in one place

**Comment by MrHobbes74** [comment]
Not planning, but new apis for other vehicles may result in new source files. Let me promise that in a next PR, after decoupling I will create an EV folder for all api’s and a generic EV worker.

**Comment by gizmocuz** [comment]
Thanks!!

**Comment by MrHobbes74** [comment]
Thanks for integrating!

**Comment by ebretteville** [comment]
Hye @MrHobbes74 , just updated today to the latest beta and discovered by the official stable announcement that a tesla hardware plugin were created so I configured it right away. Everything worked fine. Only the lert indicator is missing... Any ideas why?

**Comment by MrHobbes74** [comment]
Could you please bring this to the forum? There is a tesla topic. If it is missing, you did not activate it in the devices.

**Comment by ebretteville** [comment]
> 
> 
> Could you please bring this to the forum? There is a tesla topic. If it is missing, you did not activate it in the devices.

I found the problem. Sentrymode wasn't activted, so the state device wasn't discovered.

**Comment by MrHobbes74** [comment]
Sentrymode has nothing to do with the state device. But waking up the car could have been helpful ;-)

**Pull request merged**